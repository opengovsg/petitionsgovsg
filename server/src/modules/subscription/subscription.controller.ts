import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { CreateSubscriptionReqDto, ErrorDto } from '~shared/types/api'
import { createLogger } from '@/bootstrap/logging'
import { ControllerHandler } from '@/types/response-handler'
import { SubscriptionService } from './subscription.service'

const logger = createLogger(module)

export class SubscriptionController {
  private subscriptionService: Public<SubscriptionService>

  constructor({
    subscriptionService,
  }: {
    subscriptionService: Public<SubscriptionService>
  }) {
    this.subscriptionService = subscriptionService
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @body email
   * @returns 200 with new subscription id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 403 if user is not authorized to answer question
   * @returns 500 if database error
   */
  createSubscription: ControllerHandler<
    { id: string },
    number | ErrorDto,
    CreateSubscriptionReqDto,
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: errors.array()[0].msg })
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      // Save Subscription in the database
      const subscriptionId = await this.subscriptionService.createSubscription({
        email: req.body.email,
        postId: Number(req.params.id),
      })
      return res.status(StatusCodes.OK).json(subscriptionId)
    } catch (error) {
      logger.error({
        message: 'Error while adding new subscription',
        meta: {
          function: 'createSubscription',
          userId: req.user.id,
          postId: req.params.id,
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }
  }
}
