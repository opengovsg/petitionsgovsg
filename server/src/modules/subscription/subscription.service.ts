import type { Sequelize as SequelizeType } from 'sequelize'
import { Subscription } from '~shared/types/base'
import { ModelDef } from '@/types/sequelize'

export class SubscriptionService {
  private Subscription: ModelDef<Subscription>
  private sequelize: SequelizeType

  constructor({
    Subscription,
    sequelize,
  }: {
    Subscription: ModelDef<Subscription>
    sequelize: SequelizeType
  }) {
    this.Subscription = Subscription
    this.sequelize = sequelize
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @param email subscriber's email
   * @returns id of new subscription if it is successfully created
   */
  createSubscription = async ({
    postId,
    email,
  }: Pick<Subscription, 'email' | 'postId'>): Promise<number> => {
    const subscriptionId = await this.sequelize.transaction(
      async (transaction) => {
        const subscription = await this.Subscription.create(
          {
            postId: postId,
            email: email,
          },
          { transaction },
        )
        return subscription.id
      },
    )
    return subscriptionId
  }
}
