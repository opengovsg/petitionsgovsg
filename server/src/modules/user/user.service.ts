import { User } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'
import { LoadPublicUserDto } from '~shared/types/api'
export class UserService {
  private User: ModelDef<User>

  constructor({ User }: { User: ModelDef<User> }) {
    this.User = User
  }

  loadUser = async (userId: number): Promise<LoadPublicUserDto> => {
    return this.User.findByPk(userId) as Promise<LoadPublicUserDto>
  }

  loadFullUser = async (userId: number): Promise<User> => {
    return this.User.findByPk(userId) as Promise<User>
  }

  loadUserBySgid = async (sgid: string): Promise<LoadPublicUserDto> => {
    return this.User.findOne({ where: { sgid: sgid } })
  }
}
