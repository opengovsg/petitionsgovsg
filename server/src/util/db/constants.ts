export const POST_ID = 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847'

export const INVALID_POST_ID = 'af7b6aa9-7c6e-4ea5-b536-4e55bc506840'
