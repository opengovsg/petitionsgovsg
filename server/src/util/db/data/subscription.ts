export const mockSubscription = {
  id: 1,
  email: 'test@email.com',
  postId: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}
