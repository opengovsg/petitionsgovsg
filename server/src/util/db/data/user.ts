import { UserAuthType } from '~shared/types/api'

export const mockUser = {
  id: 'sgid',
  fullname: 'John Doe',
  type: UserAuthType.Public,
}
