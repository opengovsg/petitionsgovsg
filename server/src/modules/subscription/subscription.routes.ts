import { AuthMiddleware } from '../auth/auth.middleware'
import express from 'express'
import { body } from 'express-validator'
import { SubscriptionController } from './subscription.controller'
import { limiter } from '@/middleware/limiter'

export const routeSubscriptions = ({
  controller,
  authMiddleware,
}: {
  controller: SubscriptionController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()
  const { authenticate } = authMiddleware

  /**
   * Create a subscription attached to a post
   * @route   POST /api/subscriptions/:id
   * @returns 200 with new subscriptions id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.post(
    '/:id',
    [
      authenticate,
      limiter,
      body('email', 'email should be a valid format').isString(),
    ],
    controller.createSubscription,
  )
  return router
}
