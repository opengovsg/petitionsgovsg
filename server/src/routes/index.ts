import express from 'express'
import { AddresseeController } from '@/modules/addressee/addressee.controller'
import { routeAddressees } from '@/modules/addressee/addressee.routes'
import { AuthController } from '@/modules/auth/auth.controller'
import { AuthMiddleware } from '@/modules/auth/auth.middleware'
import { routeAuth } from '@/modules/auth/auth.routes'
import { EnvController } from '@/modules/environment/env.controller'
import { routeEnv } from '@/modules/environment/env.routes'
import { PostController } from '@/modules/post/post.controller'
import { routePosts } from '@/modules/post/post.routes'
import { SignatureController } from '@/modules/signatures/signature.controller'
import { routeSignatures } from '@/modules/signatures/signature.routes'
import { SubscriptionController } from '@/modules/subscription/subscription.controller'
import { routeSubscriptions } from '@/modules/subscription/subscription.routes'

type ApiRouterOptions = {
  signature: {
    controller: SignatureController
    authMiddleware: AuthMiddleware
  }
  env: EnvController
  auth: {
    controller: AuthController
    authMiddleware: AuthMiddleware
  }
  post: {
    controller: PostController
    authMiddleware: AuthMiddleware
  }
  addressee: {
    controller: AddresseeController
  }
  subscription: {
    controller: SubscriptionController
    authMiddleware: AuthMiddleware
  }
}

export const api = (options: ApiRouterOptions): express.Router => {
  const router = express.Router()

  router.use('/auth', routeAuth(options.auth))
  router.use('/posts', routePosts(options.post))
  router.use('/environment', routeEnv({ controller: options.env }))
  router.use('/posts/signatures', routeSignatures(options.signature))
  router.use('/addressees', routeAddressees(options.addressee))
  router.use('/subscriptions', routeSubscriptions(options.subscription))

  return router
}
