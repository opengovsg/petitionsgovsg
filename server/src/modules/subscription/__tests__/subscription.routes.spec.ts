import supertest from 'supertest'
import express from 'express'
import { StatusCodes } from 'http-status-codes'
import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '@/util/db/jest-db'
import { routeSubscriptions } from '../subscription.routes'
import { SubscriptionService } from '../subscription.service'
import { SubscriptionController } from '../subscription.controller'
import { ControllerHandler } from '@/types/response-handler'
import { mockUser, mockUserJWT } from '@/util/db/data/user'
import { POST_ID } from '@/util/db/constants'
import { mockSubscription } from '@/util/db/data/subscription'

describe('/subscriptions', () => {
  let db: SequelizeWithModels

  // Set up auth middleware to inject user
  let authUser: Express.User | undefined
  let jwt: { jwt: string } | undefined
  const authenticate: ControllerHandler = (req, _res, next) => {
    req.user = authUser
    req['cookies'] = jwt
    next()
  }

  const app = express()
  const request = supertest(app)
  app.use(express.json())

  beforeAll(async () => {
    db = await createTestDatabase()

    const subscriptionService = new SubscriptionService({
      Subscription: db.Subscription,
      sequelize: db.sequelize,
    })
    const controller = new SubscriptionController({
      subscriptionService,
    })
    app.use(
      routeSubscriptions({
        controller,
        authMiddleware: { authenticate },
      }),
    )
  })

  beforeEach(async () => {
    authUser = mockUser
    jwt = mockUserJWT
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('Post /:id', () => {
    it('returns OK on valid submission', async () => {
      const response = await request
        .post(`/${POST_ID}`)
        .send({ email: mockSubscription.email })

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toMatchSnapshot()
    })
    it('returns UNAUTHORIZED on no user', async () => {
      authUser = undefined
      jwt = undefined
      const response = await request
        .post(`/${POST_ID}`)
        .send({ email: mockSubscription.email })

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toMatchSnapshot()
    })
    it('returns BAD REQUEST on bad email', async () => {
      const response = await request.post(`/${POST_ID}`).send({ email: null })

      expect(response.status).toEqual(StatusCodes.BAD_REQUEST)
      expect(response.body).toMatchSnapshot()
    })
  })
})
