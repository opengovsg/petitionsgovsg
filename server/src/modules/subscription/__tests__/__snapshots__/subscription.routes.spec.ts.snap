// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/subscriptions Post /:id returns BAD REQUEST on bad email 1`] = `
Object {
  "message": "email should be a valid format",
}
`;

exports[`/subscriptions Post /:id returns OK on valid submission 1`] = `1`;

exports[`/subscriptions Post /:id returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;
